### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by shimondoodkin**
Windows support

### CODE REVIEW

**Comment by mhassan1** [comment]
Thank you for the PR! I really appreciate all the work that you put into this, but I feel uncomfortable about distributing an unofficial Redis binary from GitHub. Maybe we can force the consumer of this library to provide the URL to the binary via configuration so that it is clear that they are using it at their own risk. What do you think?

File: src/util/RedisBinary.ts [hunk 1]
```diff
@@ -40,8 +40,8 @@ export default class RedisBinary {
 
       log(`RedisBinary: found system binary path at "${systemBinary}"`);
       binaryPath = systemBinary;
-    } catch (err) {
-      log(`RedisBinary: can't find system binary at "${systemBinary}".\n${err.message}`);
+    } catch (err: any) {
```
Comment (mhassan1) [inline src/util/RedisBinary.ts:43]: can we use `Error` instead of `any`?

**Comment by mhassan1** [comment]
please run `yarn` and commit the changes to `yarn.lock`

**Comment by mhassan1** [comment]
please add `windows` to the travis configuration. it should look like this (i think):
```yml
language: node_js
node_js:
  - lts/*
  - node
jobs:
  - os: linux
    arch: arm64
  - os: windows
env:
  - YARN_GPG=no
cache:
  directories:
    - .yarn/cache
```

File: src/util/RedisBinary.ts (hunk 1 above)
Comment (shimondoodkin) [reply src/util/RedisBinary.ts:43]: linter says:

  Catch clause variable type annotation must be 'any' or 'unknown' if specified.ts(1196)

File: src/util/RedisBinaryDownload.ts [hunk 2]
```diff
@@ -315,7 +372,7 @@ export default class RedisBinaryDownload {
     try {
       await promisify(fs.lstat)(location);
       return true;
-    } catch (e) {
+    } catch (e: any) {
```
[the remark this answers is not in the document]
Comment (shimondoodkin) [reply src/util/RedisBinaryDownload.ts:375]: same here

File: src/util/RedisBinaryDownloadUrl.ts [hunk 3]
```diff
@@ -22,15 +22,19 @@ export default class RedisBinaryDownloadUrl {
...
-    log(`Using "${archive}" as the Archive String`);
-
     const downloadUrl = resolveConfig('DOWNLOAD_URL');
     if (downloadUrl) {
       log(`Using "${downloadUrl}" as the Download-URL`);
       return downloadUrl;
     }
 
+    if (process.platform === 'win32') {
+      return 'https://github.com/ServiceStack/redis-windows/raw/master/downloads/redis-latest.zip';
```
[the remark this answers is not in the document]
Comment (shimondoodkin) [reply src/util/RedisBinaryDownloadUrl.ts:32]: there is no other option, for windows

File: README.md [hunk 4]
```diff
@@ -50,7 +50,10 @@ On install, this package auto-downloads and compiles version `stable` of the `re
 - Typescript: 3.8+ (if used)
 - `make`
 
-NOTE: Windows is not officially supported by this library, since it is not officially supported by Redis. PRs welcome!
+### Windows:
+
+
+Although Windows is not officially supported by Redis, but there is a downloadable version of [redis: 3.0.503 (June 28, 2016) for win64 by microsoft](https://github.com/ServiceStack/redis-windows#option-3-running-microsofts-native-port-of-redis). This is an older version. but it will work for most of the simple tasks like basic storage and pubsub. Probably what is missing or old is lua scripting.
```
[the remark this answers is not in the document]
Comment (shimondoodkin) [reply README.md]: see new text

File: src/util/RedisBinaryDownloadUrl.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mhassan1) [reply src/util/RedisBinaryDownloadUrl.ts:32]: understood. i see you have added it to the documentation, which works for me. thanks.

File: src/util/RedisBinaryDownload.ts [hunk 5]
```diff
@@ -202,62 +219,87 @@ export default class RedisBinaryDownload {
...
-                    "Try to use different version 'new RedisMemoryServer({ binary: { version: 'X.Y.Z' } })'\n"
-                )
-              );
+
+      const get = () => {
+        https
+          .get(httpOptions as any, (response) => {
+            // "as any" because otherwise the "agent" wouldnt match
+            if (response.statusCode != 200) {
+              if (response.statusCode === 301 || response.statusCode === 302) {
```
Comment (mhassan1) [inline src/util/RedisBinaryDownload.ts:228]: why did we need to add all this redirect logic? was it to support the redirect from `github.com` to `raw.githubusercontent.com`? it seems like we could just use the `raw.githubusercontent.com` directly, and then we shouldn't need to add all this code.

**Comment by mhassan1** [comment]
I'm trying to figure out what this build failure is all about: https://app.travis-ci.com/github/mhassan1/redis-memory-server/jobs/538797915

Then, we will need to get a passing build on Windows. Right now, it's complaining about line endings: https://app.travis-ci.com/github/mhassan1/redis-memory-server/jobs/538797917

**Comment by mhassan1** [comment]
Regarding Linux test failures: there may be some interaction between tests running in parallel. Can you try adding `--runInBand` to the `coverage` script in `package.json`?

**Comment by mhassan1** [comment]
I have resolved all CI issues on a local branch: https://app.travis-ci.com/github/mhassan1/redis-memory-server/builds/238300548

I've created a PR for you to merge the local branch into your `windows` branch: https://github.com/shimondoodkin/redis-memory-server/pull/1

Once that PR is merged, we should see CI pass here, and then we can merge and release this PR.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "redis-memory-server",
-  "version": "0.3.1",
+  "version": "0.4.0",
```
Comment (mhassan1) [inline package.json:3]: please remove this change. i will bump the version as part of the release process.

File: src/util/RedisBinaryDownload.ts (hunk 5 above)
Comment (shimondoodkin) [reply src/util/RedisBinaryDownload.ts:228]: yes, GitHub serves files from aws, it generates a temporary authorized URL first then redirects to it. It's good enough.

Comment (mhassan1) [reply src/util/RedisBinaryDownload.ts:228]: can you give me an example of a URL that requires this code?

Comment (shimondoodkin) [reply src/util/RedisBinaryDownload.ts:228]: i see it is different now, sometime ago github had this aws redirect. when I had a redirect i solved it.

now i see it gives file from 

https://raw.githubusercontent.com/ServiceStack/redis-windows/master/downloads/redis-latest.zip

**Comment by shimondoodkin** [comment]
the CRs are from
git config auto.crlf true

need to set it to false
git config auto.crlf false

where is the travis build script?

**Comment by shimondoodkin** [comment]
I just saw your pull request, I merged it. Thanks

**Review by mhassan1 — APPROVED** [review]

**Review by mhassan1 — APPROVED** [review]
all checks passing! thanks again for the contribution!

**Pull request merged**